### ISSUE TITLE
Q: How do you ensure the HEC token value is obfuscated?

### ISSUE THREAD

**Issue by pranavmarla** [splunk/splunk-connect-for-kubernetes#527]
This is a question, not a bug report. **I was hoping you could explain the way you've (I THINK) used Kubernetes Secrets to obfuscate the HEC token, because I'm having trouble understanding how it's meant to work.** Details below:

###############

I am concerned that, after the Splunk HEC token is passed to this Helm chart via the `splunk.hec.token` value, the HEC token will be visible (in plaintext) to anyone who runs `helm get values <CHART_NAME>`. Initially, I assumed that you solved this problem using a Kubernetes Secret but, now that I examine your code, I don't understand what it's accomplishing, or how it's meant to work. 

As far as I can tell, these are the steps being performed in this Helm chart:
1. Take in HEC token via Helm value named `splunk.hec.token` ([link](https://github.com/splunk/splunk-connect-for-kubernetes/blob/develop/helm-chart/splunk-connect-for-kubernetes/charts/splunk-kubernetes-logging/values.yaml#L62))
2. Create a Kubernetes secret to store that token, named `splunk_hec_token` ([link](https://github.com/splunk/splunk-connect-for-kubernetes/blob/develop/helm-chart/splunk-connect-for-kubernetes/charts/splunk-kubernetes-logging/templates/secret.yaml#L13-L14))
3. Create an environment variable to store that token, named `SPLUNK_HEC_TOKEN` ([link](https://github.com/splunk/splunk-connect-for-kubernetes/blob/develop/helm-chart/splunk-connect-for-kubernetes/charts/splunk-kubernetes-logging/templates/daemonset.yaml#L69-L73))
4. Subsequently, whenever you need to use the HEC token, you reference the `SPLUNK_HEC_TOKEN` environment variable ([link](https://github.com/splunk/splunk-connect-for-kubernetes/blob/develop/helm-chart/splunk-connect-for-kubernetes/charts/splunk-kubernetes-logging/templates/configMap.yaml#L335))

**What's confusing me about this approach to obfuscating the HEC token is:**
1. Isn't the original Helm value containing the HEC token (`splunk.hec.token`) still visible in plaintext to anyone who runs `helm get values <CHART_NAME>`?
2. If the `splunk.hec.token` Helm value is still visible in plaintext, then what is the point of creating the Kubernetes Secret and environment variable? 
In other words, if we're not actually being more secure -- if we're not actually obfuscating the HEC token contained within the `splunk.hec.token` Helm value -- then why even bother creating the Kubernetes Secret and environment variable? What is it actually accomplishing here?

**Comment by vinzent** [comment]
> to anyone who runs helm get values <CHART_NAME>

The helm state is stored as Kubernetes Secret too. Only someone with rights to retrieve the Secret is able to access the the Helm values. So `helm get values <chart-name>` can only be run by somebody that is permissioned to access Secrets on that namespace anyway.

**Comment by pranavmarla** [comment]
@vinzent Interesting, I hadn't heard of that. But then, if that's true, that leads back to my original question: Why bother creating a Kubernetes secret at all? 
If a Helm value is really just as secure as a Kubernetes secret, why not just directly reference the Helm value in the Fluentd config -- why go to all the trouble of creating a Kubernetes secret based off of the Helm value, and then an env variable based off of the Kubernetes secret??

**Comment by vinzent** [comment]
The helm value can't be read by fluentd because the fluentd config file format ist not helm values. The helm state stored in the secret is for Helm to know the state. But this is really a Helm topic, not splunk-connect. 

But anyway, fluentd by default spits out its complete config on startup including the HEC token. Actually anybody with access to the logs of the pod has access to the HEC credential tool.

**Comment by pranavmarla** [comment]
@vinzent Ah, I think I get it now -- the reason you create the Kubernetes secret + env variable is to get it into a form (env variable) that can be consumed by Fluentd without revealing its value (HEC token) in plaintext when Fluentd dumps its config on startup (I'm assuming here that Fluentd does not print out the actual value of env variables in its config).

In other words, although we could directly reference the Helm value in the Fluentd config (although Fluentd doesn't directly consume (or know about) Helm values, Helm will substitute the value beforehand so that, from Fluentd's perspective, it's as if you hardcoded the Helm value (HEC token) in plaintext in the config -- see example [here](https://github.com/splunk/splunk-connect-for-kubernetes/blob/develop/helm-chart/splunk-connect-for-kubernetes/charts/splunk-kubernetes-logging/templates/configMap.yaml#L337)), the HEC token would now be printed in plaintext when Fluentd dumps its config on startup.

Thanks for explaining! Assuming I understood you correctly, feel free to close this ticket.

**Comment by rockb1017** [comment]
Thank you @vinzent for your explanation! closing
